ath79: add YunCore TFTP image generation helper

CodeFetch opened the pull request:
YunCore routers based on the AP147 reference board need a prepended U-Boot command for accepting TFTP images.
This command changes the bootcmd to accept OpenWrt's flash layout and boot at the stock firmware address if this fails
and writes the firmware image to flash afterwards.

Signed-off-by: Vincent Wiemann <vincent.wiemann@ironai.com>

CodeFetch committed:
ath79: add support for YunCore TFTP image generation

YunCore routers based on the AP147 reference board need a
prepended U-Boot command for accepting TFTP images.
This command changes the bootcmd to accept OpenWrt's flash
layout and boot at the stock firmware address if this fails.
Afterwards it writes the firmware image to flash.

This commit adds the helper script and enables TFTP image
generation for the YunCore A770.

Signed-off-by: Vincent Wiemann <vincent.wiemann@ironai.com>

adrianschmutzler commented on target/linux/ath79/image/common-yuncore.mk:
```diff
@@ -0,0 +1,11 @@
+		echo -n -e "YUNCOREsetenv bootcmd \"bootm 0x9f050000 || bootm 0x9fe80000\" &&" \
+			"saveenv && erase 0x9f050000 +0xfa0000 && cp.b 0x800600c0 0x9f050000 0xfa0000" | \
+		dd bs=192 count=1 conv=sync; \
+		  dd if=$@; \
```
Mixed indent here by intention?

adrianschmutzler commented on target/linux/ath79/image/common-yuncore.mk:
```diff
@@ -0,0 +1,11 @@
+		dd bs=192 count=1 conv=sync; \
+		  dd if=$@; \
+	) > $@.new
+  mv $@.new $@
```
Spaces for indent by intention?

CodeFetch replied:
Nope. Did the changes with Github UI which replaces tabs with spaces. Going to fix it.

CodeFetch commented:
Also added TFTP image generation for YunCore A770 as I've just realized that I've one lying around here and it has the same U-Boot functionality.

adrianschmutzler commented:
I would either make it two commits (image generation + device changes) or rephrase the existing commit message then ... ;-)

CodeFetch commented:
@adrianschmutzler Changed commit message...

Neustradamus commented:
It is for?
- YunCore A770
- YunCore AP80Q
- YunCore AP90Q
- YunCore CPE830
- YunCore CPE830D
- YunCore CPE870
- YunCore T830
- YunCore SR3200
- YunCore XD3200

CodeFetch commented:
No. Some of the devices you mentioned are AP152-based and I don't have a clue about them. Likely they have the same functionality, but I don't know.

grosjo commented:
Once again, 830 or 830D, this is a quesiton of leds, who cares....

pepe2k commented:
@CodeFetch A770 I have doesn't have this feature... can you send me U-Boot dump from yours? Was that confirmed on any other QCA953x based YunCore devices/models?

Cheers,
Piotr

CodeFetch commented:
@pepe2k Maybe only newer devices have this feature. I've also confirmed this with T740.
Here's the U-Boot dump: https://0x0.st/zVIU.bin

pepe2k commented:
@CodeFetch

Confirmed, it probably applies only to a recently sold YunCore products, with updated U-Boot.
My A770 has `U-Boot 1.1.4 (Oct 24 2017 - 13:22:56)`.

I would be also very careful with erasing and offsets - what in case someone gets cheaper version of the board with just 8 MB of flash? Could you show me log from serial console from whole TFTP update procedure performed on A770?

And one more comment from me... please, drop that AP147 reference and suffix from commit message and recipe name. It's just confusing and if it turns out that YunCore devices with other WiSOCs also have this feature... we would need to rename it :)

Cheers,
Piotr

CodeFetch commented:
@pepe2k Serial log:

```
U-Boot 1.1.4 (Apr 11 2019 - 14:56:21)

ap147 - Honey Bee 2.0DRAM:  
sri
Honey Bee 2.0
ath_ddr_initial_config(195): (16bit) ddr2 init
ath_sys_frequency: cpu 650 ddr 575 ahb 216
tap = 0x00000003
Tap (low, high) = (0x7, 0x24)
Tap values = (0x15, 0x15, 0x15, 0x15)
128 MB
Top of RAM usable for U-Boot at: 88000000
Reserving 202k for U-Boot at: 87fcc000
Reserving 192k for malloc() at: 87f9c000
Reserving 44 Bytes for Board Info at: 87f9bfd4
Reserving 36 Bytes for Global Data at: 87f9bfb0
Reserving 128k for boot params() at: 87f7bfb0
Stack Pointer at: 87f7bf98
Now running in RAM - U-Boot at: 87fcc000
Flash Manuf Id 0xef, DeviceId0 0x40, DeviceId1 0x18
flash size 16MB, sector count = 256
Flash: 16 MB
Power up PLL with outdiv = 0 then switch to 3
In:    serial
Out:   serial
Err:   serial
Net:   ath_gmac_enet_initialize...
No valid address in Flash. Using fixed address
No valid address in Flash. Using fixed address
ath_gmac_enet_initialize: reset mask:c02200 
Honey Bee ---->S27 PHY*
S27 reg init
: cfg1 0x800c0000 cfg2 0x7114
eth0: 00:03:7f:09:0b:ad
athrs27_phy_setup ATHR_PHY_CONTROL 4 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 4 :10
eth0 up
Honey Bee ---->  MAC 1 S27 PHY *
S27 reg init
ATHRS27: resetting s27
ATHRS27: s27 reset done
: cfg1 0x800c0000 cfg2 0x7214
eth1: 00:03:7f:09:0b:ad
athrs27_phy_setup ATHR_PHY_CONTROL 0 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 0 :10
athrs27_phy_setup ATHR_PHY_CONTROL 1 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 1 :10
athrs27_phy_setup ATHR_PHY_CONTROL 2 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 2 :10
athrs27_phy_setup ATHR_PHY_CONTROL 3 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 3 :10
eth1 up
eth0, eth1
Setting 0x181162c0 to 0x811a100
Warning: reset button has been pushed!!!!!!
checking the reset button cmd in firmware
download firmware: tftp 0x80060000 upgrade.bin 
Trying eth0
eth0 link down
FAIL
Trying eth1
eth1 link down
FAIL
Using eth0 device
TFTP from server 192.168.0.141; our IP address is 192.168.0.144
Filename 'upgrade.bin'.
Load address: 0x80060000
Loading: Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out
T Tx Timed out

Retry count exceeded; starting again
Trying eth1
enet1 port3 up
dup 1 speed 1000
Using eth1 device
TFTP from server 192.168.0.141; our IP address is 192.168.0.144
Filename 'upgrade.bin'.
Load address: 0x80060000
Loading: #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
done
Bytes transferred = 4325841 (4201d1 hex)
firmware resetcmd string: YUNCOREsetenv bootcmd "bootm 0x9f050000 || bootm 0x9fe80000" && saveenv && erase 0x9f050000 +0xfa0000 && cp.b 0x800600c0 0x9f050000 0xfa0000 
Please wait......
Saving Environment to Flash...
Protect off 9F040000 ... 9F04FFFF
Un-Protecting sectors 4..4 in bank 1
Un-Protected 1 sectors
Erasing Flash...Erasing flash... 
First 0x4 last 0x4 sector size 0x10000                                                                                                                                   4
Erased 1 sectors
Writing to Flash... write addr: 9f040000
done
Protecting sectors 4..4 in bank 1
Protected 1 sectors
Erasing flash... 
First 0x5 last 0xfe sector size 0x10000                                                                                                                                254
Erased 250 sectors
Copy to Flash... write addr: 9f050000
done

NOTIFICATION: firmware resetcmd FINISHED
Rebooting......

U-Boot 1.1.4 (Apr 11 2019 - 14:56:21)

ap147 - Honey Bee 2.0DRAM:  
sri
Honey Bee 2.0
ath_ddr_initial_config(195): (16bit) ddr2 init
ath_sys_frequency: cpu 650 ddr 597 ahb 216
tap = 0x00000003
Tap (low, high) = (0x7, 0x23)
Tap values = (0x15, 0x15, 0x15, 0x15)
128 MB
Top of RAM usable for U-Boot at: 88000000
Reserving 202k for U-Boot at: 87fcc000
Reserving 192k for malloc() at: 87f9c000
Reserving 44 Bytes for Board Info at: 87f9bfd4
Reserving 36 Bytes for Global Data at: 87f9bfb0
Reserving 128k for boot params() at: 87f7bfb0
Stack Pointer at: 87f7bf98
Now running in RAM - U-Boot at: 87fcc000
Flash Manuf Id 0xef, DeviceId0 0x40, DeviceId1 0x18
flash size 16MB, sector count = 256
Flash: 16 MB
Power up PLL with outdiv = 0 then switch to 3
In:    serial
Out:   serial
Err:   serial
Net:   ath_gmac_enet_initialize...
No valid address in Flash. Using fixed address
No valid address in Flash. Using fixed address
ath_gmac_enet_initialize: reset mask:c02200 
Honey Bee ---->S27 PHY*
S27 reg init
: cfg1 0x800c0000 cfg2 0x7114
eth0: 00:03:7f:09:0b:ad
athrs27_phy_setup ATHR_PHY_CONTROL 4 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 4 :10
eth0 up
Honey Bee ---->  MAC 1 S27 PHY *
S27 reg init
ATHRS27: resetting s27
ATHRS27: s27 reset done
: cfg1 0x800c0000 cfg2 0x7214
eth1: 00:03:7f:09:0b:ad
athrs27_phy_setup ATHR_PHY_CONTROL 0 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 0 :10
athrs27_phy_setup ATHR_PHY_CONTROL 1 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 1 :10
athrs27_phy_setup ATHR_PHY_CONTROL 2 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 2 :10
athrs27_phy_setup ATHR_PHY_CONTROL 3 :1000
athrs27_phy_setup ATHR_PHY_SPEC_STAUS 3 :10
eth1 up
eth0, eth1
Setting 0x181162c0 to 0x811a100
Hit any key to stop autoboot:  0 
## Booting image at 9f050000 ...
   Image Name:   MIPS OpenWrt Linux-4.19.66
   Created:      2019-08-15  12:25:51 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    1888730 Bytes =  1.8 MB
   Load Address: 80060000
   Entry Point:  80060000
   Verifying Checksum at 0x9f050040 ...OK
   Uncompressing Kernel Image ... OK
No initrd
## Transferring control to Linux (at address 80060000) ...
## Giving linux memsize in bytes, 134217728

Starting kernel ...

[    0.000000] Linux version 4.19.66 (builder@buildhost) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r10766-3032bf7f89)) #0 Thu Aug 15 12:25:51 2019
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019374 (MIPS 24Kc)
[    0.000000] MIPS: machine is YunCore A770
[    0.000000] SoC: Qualcomm Atheros QCA9533 ver 2 rev 0
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] random: get_random_bytes called from start_kernel+0x98/0x4b0 with crng_init=0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200n8 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 122164K/131072K available (4502K kernel code, 186K rwdata, 1052K rodata, 1228K init, 214K bss, 8908K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 650.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 5880801374 ns
[    0.000008] sched_clock: 32 bits at 325MHz, resolution 3ns, wraps every 6607641598ns
[    0.008237] Calibrating delay loop... 432.53 BogoMIPS (lpj=2162688)
[    0.074849] pid_max: default: 32768 minimum: 301
[    0.079977] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.086947] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.098969] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.109389] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.115961] pinctrl core: initialized pinctrl subsystem
[    0.124070] NET: Registered protocol family 16
[    0.135964] PCI host bridge /ahb/pcie-controller@180c0000 ranges:
[    0.142461]  MEM 0x0000000010000000..0x0000000013ffffff
[    0.147960]   IO 0x0000000000000000..0x0000000000000000
[    0.172691] PCI host bridge to bus 0000:00
[    0.177080] pci_bus 0000:00: root bus resource [mem 0x10000000-0x13ffffff]
[    0.184319] pci_bus 0000:00: root bus resource [io  0x0000]
[    0.190211] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.197367] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.206989] pci 0000:00:00.0: BAR 0: assigned [mem 0x10000000-0x101fffff 64bit]
[    0.214716] pci 0000:00:00.0: BAR 6: assigned [mem 0x10200000-0x1020ffff pref]
[    0.225924] clocksource: Switched to clocksource MIPS
[    0.232452] NET: Registered protocol family 2
[    0.237981] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes)
[    0.246154] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.253508] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.260236] TCP: Hash tables configured (established 1024 bind 1024)
[    0.267100] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.273266] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.280247] NET: Registered protocol family 1
[    0.288028] Crashlog allocated RAM at address 0x3f00000
[    0.294967] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.308702] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.314849] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.339247] io scheduler noop registered
[    0.343387] io scheduler deadline registered (default)
[    0.351627] pinctrl-single 1804002c.pinmux: 576 pins, size 72
[    0.358709] Serial: 8250/16550 driver, 16 ports, IRQ sharing enabled
[    0.367822] console [ttyS0] disabled
[    0.371623] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 9, base_baud = 1562500) is a 16550A
[    0.380780] console [ttyS0] enabled
[    0.380780] console [ttyS0] enabled
[    0.388349] bootconsole [early0] disabled
[    0.388349] bootconsole [early0] disabled
[    0.415834] m25p80 spi0.0: w25q128 (16384 Kbytes)
[    0.420811] 4 fixed-partitions partitions found on MTD device spi0.0
[    0.427391] Creating 4 MTD partitions on "spi0.0":
[    0.432367] 0x000000000000-0x000000040000 : "u-boot"
[    0.438378] 0x000000040000-0x000000050000 : "u-boot-env"
[    0.444603] 0x000000050000-0x000000ff0000 : "firmware"
[    0.453157] 2 uimage-fw partitions found on MTD device firmware
[    0.459333] Creating 2 MTD partitions on "firmware":
[    0.464464] 0x000000000000-0x0000001d0000 : "kernel"
[    0.470393] 0x0000001d0000-0x000000fa0000 : "rootfs"
[    0.476316] mtd: device 4 (rootfs) set to be root filesystem
[    0.483710] 1 squashfs-split partitions found on MTD device rootfs
[    0.490171] 0x000000420000-0x000000fa0000 : "rootfs_data"
[    0.496606] 0x000000ff0000-0x000001000000 : "art"
[    0.503295] libphy: Fixed MDIO Bus: probed
[    0.857018] ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
[    1.536322] libphy: ag71xx_mdio: probed
[    1.541712] libphy: ar8xxx-mdio: probed
[    1.554573] switch0: Atheros AR8229 rev. 1 switch registered on mdio-bus.0
[    1.607464] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.617441] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    1.626451] NET: Registered protocol family 10
[    1.637102] Segment Routing with IPv6
[    1.640993] NET: Registered protocol family 17
[    1.645652] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    1.659091] 8021q: 802.1Q VLAN Support v1.8
[    1.998197] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd042, driver=Generic PHY]
[    2.008951] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    2.016170] hctosys: unable to open rtc device (rtc0)
[    2.028808] VFS: Mounted root (squashfs filesystem) readonly on device 31:4.
[    2.042814] Freeing unused kernel memory: 1228K
[    2.047537] This architecture does not have kernel memory protection.
[    2.054182] Run /sbin/init as init process
[    2.565933] random: fast init done
[    2.996487] init: Console is alive
[    3.000240] init: - watchdog -
[    4.541197] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    4.712521] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    4.730496] init: - preinit -
[    6.250373] random: jshn: uninitialized urandom read (4 bytes read)
[    6.531357] random: jshn: uninitialized urandom read (4 bytes read)
[    6.599514] random: jshn: uninitialized urandom read (4 bytes read)
[    6.861856] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    7.877107] eth0: link up (1000Mbps/Full duplex)
[    7.881912] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   10.116509] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[   10.152784] urandom-seed: Seed file not found (/etc/urandom.seed)
[   10.255047] eth0: link down
[   10.270830] procd: - early -
[   10.273920] procd: - watchdog -
[   10.912501] procd: - watchdog -
[   10.916128] procd: - ubus -
[   10.944221] urandom_read: 5 callbacks suppressed
[   10.944230] random: ubusd: uninitialized urandom read (4 bytes read)
[   10.969952] random: ubusd: uninitialized urandom read (4 bytes read)
[   10.978022] procd: - init -
Please press Enter to activate this console.
[   11.775208] kmodloader: loading kernel modules from /etc/modules.d/*
[   11.820792] Loading modules backported from Linux version v5.2-rc7-0-g6fbc7275c7a9
[   11.828680] Backport generated by backports.git v5.2-rc7-1-0-g021a6ba1
[   11.897024] xt_time: kernel timezone is -0000
[   12.118683] urngd: v1.0.0 started.
[   12.168559] PPP generic driver version 2.4.2
[   12.175597] NET: Registered protocol family 24
[   12.259735] ath10k 4.19 driver, optimized for CT firmware, probing pci device: 0x50.
[   12.287015] ath10k_pci 0000:00:00.0: enabling device (0000 -> 0002)
[   12.293643] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   12.441766] random: crng init done
[   12.792709] firmware ath10k!fwcfg-pci-0000:00:00.0.txt: firmware_loading_store: map pages failed
[   13.015217] firmware ath10k!pre-cal-pci-0000:00:00.0.bin: firmware_loading_store: map pages failed
[   13.422892] firmware ath10k!QCA9887!hw1.0!ct-firmware-5.bin: firmware_loading_store: map pages failed
[   13.643091] firmware ath10k!QCA9887!hw1.0!ct-firmware-2.bin: firmware_loading_store: map pages failed
[   13.864401] firmware ath10k!QCA9887!hw1.0!firmware-6.bin: firmware_loading_store: map pages failed
[   14.087170] firmware ath10k!QCA9887!hw1.0!firmware-5.bin: firmware_loading_store: map pages failed
[   14.307947] firmware ath10k!QCA9887!hw1.0!firmware-4.bin: firmware_loading_store: map pages failed
[   14.530935] firmware ath10k!QCA9887!hw1.0!firmware-3.bin: firmware_loading_store: map pages failed
[   14.544734] ath10k_pci 0000:00:00.0: qca9887 hw1.0 target 0x4100016d chip_id 0x004000ff sub 0000:0000
[   14.554330] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   14.567974] ath10k_pci 0000:00:00.0: firmware ver 10.1-ct-87-__fW-022-b7d05fdf api 2 features wmi-10.x,has-wmi-mgmt-tx,mfp,txstatus-noack,wmi-10.x-CT,ratemask-CT,txrate-CT,get-temp-CT,tx-rc-CT,cust-stats-CT,retry-gt2-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT crc32 1a63a0ad
[   14.826225] firmware ath10k!QCA9887!hw1.0!board-2.bin: firmware_loading_store: map pages failed
[   14.835758] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id N/A crc32 546cca0d
[   15.827860] ath10k_pci 0000:00:00.0: 10.1 wmi init: vdevs: 16  peers: 127  tid: 256
[   15.841825] ath10k_pci 0000:00:00.0: wmi print 'P 128 V 8 T 410'
[   15.848337] ath10k_pci 0000:00:00.0: wmi print 'msdu-desc: 1424  sw-crypt: 0 ct-sta: 0'
[   15.856645] ath10k_pci 0000:00:00.0: wmi print 'alloc rem: 21496 iram: 12132'
[   15.907253] ath10k_pci 0000:00:00.0: htt-ver 2.1 wmi-op 2 htt-op 2 cal file max-sta 128 raw 0 hwcrypto 1
[   15.920429] ath10k_pci 0000:00:00.0: NOTE:  Firmware DBGLOG output disabled in debug_mask: 0x10000000
[   16.167125] ieee80211 phy1: Atheros AR9531 Rev:2 mem=0xb8100000, irq=13
[   16.249624] kmodloader: done loading kernel modules from /etc/modules.d/*
[   64.723682] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   64.735307] jffs2_build_filesystem(): unlocking the mtd device... 
[   64.735377] done.
[   64.743822] jffs2_build_filesystem(): erasing all blocks after the end marker... 
[   68.269556] eth0: link up (1000Mbps/Full duplex)
[   68.287278] br-lan: port 1(eth0) entered blocking state
[   68.292686] br-lan: port 1(eth0) entered disabled state
[   68.298475] device eth0 entered promiscuous mode
[   68.343242] br-lan: port 1(eth0) entered blocking state
[   68.348707] br-lan: port 1(eth0) entered forwarding state
[   68.354491] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   68.475524] IPv6: ADDRCONF(NETDEV_UP): eth1: link is not ready
[   69.315979] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[  107.297198] done.
[  107.299338] jffs2: notice: (1376) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[  107.492438] overlayfs: upper fs does not support tmpfile.
```

CodeFetch commented:
@pepe2k 

> I would be also very careful with erasing and offsets - what in case someone gets cheaper version of the board with just 8 MB of flash?

Are you sure that such versions exist? I can ask YunCore if not. The problem here is that this U-Boot version does not have any commands for doing "if-like" checks or setting environment variables with the result of an arithmetic operation. If they'd have such features I could just subtract the size of the TFTP header from firmware size and don't erase/overwrite the maximum size one image could have. I know that, but if there are no 8 MB devices, I wouldn't care.

> And one more comment from me... please, drop that AP147 reference and suffix from commit message and recipe name. It's just confusing and if it turns out that YunCore devices with other WiSOCs also have this feature... we would need to rename it :)

I would rather rename it later than having someone setting this for other APs without checking whether they have this feature. For AP147-based devices I was rather sure that they all have the same flash layout/size. I can get some AP152 devices to test, but still I'd be very careful naming this function as it is potentially dangerous.

CodeFetch commented:
@pepe2k Here are the available commands, but I don't have a clue how one could use e.g. itest for this:

```
?       - alias for 'help'
autoscr - run script from memory
base    - print or set address offset
bdinfo  - print Board Info structure
boot    - boot default, i.e., run 'bootcmd'
bootd   - boot default, i.e., run 'bootcmd'
bootelf - Boot from an ELF image in memory
bootm   - boot application image from memory
bootp   - boot image via network using BootP/TFTP protocol
bootvx  - Boot vxWorks from an ELF image
cmp     - memory compare
coninfo - print console devices and information
cp      - memory copy
crc32   - checksum calculation
dhcp    - invoke DHCP client to obtain IP/boot params
echo    - echo args to console
erase   - erase FLASH memory
ethreg  - Switch/PHY Reg rd/wr  utility
exit    - exit script
flinfo  - print FLASH memory information
go      - start application at address 'addr'
help    - print online help
iminfo  - print header information for application image
itest   - return true/false on integer compare
loop    - infinite loop on address range
mct   - simple RAM test
md      - memory display
mii     - MII utility commands
mm      - memory modify (auto-incrementing)
mtest   - simple RAM test
mw      - memory write (fill)
nfs     - boot image via network using NFS protocol
nm      - memory modify (constant address)
pci     - list and access PCI Configuration Space
ping    - send ICMP ECHO_REQUEST to network host
pll cpu-pll dither ddr-pll dither - Set to change CPU & DDR speed
pll erase
pll get
printenv- print environment variables
progmac - Set ethernet MAC addresses
progmac2 - Set ethernet MAC addresses
protect - enable or disable FLASH write protection
rarpboot- boot image via network using RARP/TFTP protocol
reset   - Perform RESET of the CPU
run     - run commands in an environment variable
saveenv - save environment variables to persistent storage
setenv  - set environment variables
sleep   - delay execution for some time
test    - minimal test like /bin/sh
tftpboot- boot image via network using TFTP protocol
version - print monitor version


```

grosjo commented on target/linux/ath79/image/common-yuncore.mk:
```diff
@@ -0,0 +1,11 @@
+define Build/yuncore-tftp-header-ap147
+	( \
+		echo -n -e "YUNCOREsetenv bootcmd \"bootm 0x9f050000 || bootm 0x9fe80000\" &&" \
+			"saveenv && erase 0x9f050000 +0xfa0000 && cp.b 0x800600c0 0x9f050000 0xfa0000" | \
```
Just replace 0xfa0000 by "$filesize"

CodeFetch replied:
It's not that easy I think as we have a 192 bytes TFTP header at the beginning and these should be substracted from the file size. A workaround could be to make the file size check in OpenWrt image generation after adding the TFTP header.

On newer U-Boots one could do `setexpr filesize $filesize - 192`, but here it isn't possible. That's why I'm not sure how to handle it.

grosjo replied:
Can you then use IMAGE_SIZEto build the line 
i.e. -e "YUNCORE...." IMAGE SIZE "...

adrianschmutzler replied:
I would not use IMAGE_SIZE, as it is subject of an ongoing discussion and might be changed/removed eventually.

pepe2k commented:
@CodeFetch 

> @pepe2k Serial log:

OK, so it seems that the flash doesn't get erased and firmware doesn't get written automatically and related commands need to be always there, am I right? What about something more generic like this (without touching old kernel area):

`YUNCOREsetenv bootcmd \"bootm 0x9f050000 || bootm 0x9fe80000\" && saveenv && erase 0x9f050000 +${filesize} && cp.b ${fileaddr} 0x9f050000 ${filesize}`

This way you just erase and write as much as was sent (assuming YunCore didn't mess with above env variables, you might need to try them without {...}).

Anyway, `erase` shouldn't allow to erase more than available flash space... but original QCA/Atheros U-Boot doesn't detect SPI NOR size, everything is static there. So, still someone with U-Boot for 16 MB version on board with just 8 MB might have problem, which won't be our problem, though.

> Are you sure that such versions exist? I can ask YunCore if not. The problem here is that this U-Boot version does not have any commands for doing "if-like" checks or setting environment variables with the result of an arithmetic operation. If they'd have such features I could just subtract the size of the TFTP header from firmware size and don't erase/overwrite the maximum size one image could have. I know that, but if there are no 8 MB devices, I wouldn't care.

As with many China vendors, YunCore can supply you with different RAM/flash configuration.
I have some of their boards with less/more flash, just not QCA953x based.

> I would rather rename it later than having someone setting this for other APs without checking whether they have this feature. For AP147-based devices I was rather sure that they all have the same flash layout/size. I can get some AP152 devices to test, but still I'd be very careful naming this function as it is potentially dangerous.

The problem is that... these devices aren't actually (exactly) based on AP147 reference design :)
AP147 doesn't refer to exact reference design, I would say, more as a "family" of designs. And it's soo confusing for regular (end) users... why don't you want to use e.g. `yuncore-tftp-header-qca953x`? It has exactly the same meaning but it's more clear for people not familiar with Qualcomm reference designs naming.
![ap147-schematic](https://user-images.githubusercontent.com/2934612/63175276-fa86f080-c043-11e9-8c4f-e094a7cfe6c2.jpg)

Cheers,
Piotr

CodeFetch commented:
@pepe2k 

> This way you just erase and write as much as was sent (assuming YunCore didn't mess with above env variables, you might need to try them without {...}).

I know, but we have the 192 bytes TFTP-header in front of the image and the OpenWrt file size check is done before adding the TFTP header. After the firmware partition there is the ART partition. What happens if a sysupgrade file has e.g. 16000K? (is this even possible? As far as I remember some space needs to be reserved for overlayfs... What is the actual value for this reservation if any?). That's why I used 0xfa0000 as lower bound.

> why don't you want to use e.g. `yuncore-tftp-header-qca953x`?

I'm fine with that. Will change it.

CodeFetch commented:
So if there is a reservation for overlayfs greater than 192 bytes, we can use $filesize without issues, but if there isn't, would it make sense to add a filesize check after adding the TFTP-header?

pepe2k commented:
@CodeFetch 

> I know, but we have the 192 bytes TFTP-header in front of the image and the OpenWrt file size check is done before adding the TFTP header.

I just saw your previous comment about that... why I'm not surprised they broke it, eh :disappointed: 

> After the firmware partition there is the ART partition. What happens if a sysupgrade file has e.g. 16000K? (is this even possible? As far as I remember some space needs to be reserved for overlayfs... What is the actual value for this reservation if any?). That's why I used 0xfa0000 as lower bound.

There is always a chance someone adds as many packages or other stuff that the whole image gets filled up to the limit.

Now I don't see better way to handle it than your initial idea, with static size and offset.

> So if there is a reservation for overlayfs greater than 192 bytes, we can use $filesize without issues, but if there isn't, would it make sense to add a filesize check after adding the TFTP-header?

Check `pad-rootfs` in `include/image-commands.mk`. AFAIR, we don't reserve space for jffs2 overlay in generated images.

> > why don't you want to use e.g. `yuncore-tftp-header-qca953x`?
> 
> I'm fine with that. Will change it.

Maybe it would be good to include also flash size in the recipe name? Like `yuncore-tftp-header-qca953x-16mb` (I'm also fine without it)?

Cheers,
Piotr

adrianschmutzler commented:
> Check pad-rootfs in include/image-commands.mk. AFAIR, we don't reserve space for jffs2 overlay in generated images.

Check out https://github.com/openwrt/openwrt/pull/2226#issuecomment-520214531

grosjo commented:
Probably add 

bootargs "console=ttyS0,115200 rootfstype=squashfs,jffs2 noinitrd"

to avoid troublesome Kuwfi / Yuncore "tweaks" (read : additional problems)

CodeFetch commented:
@grosjo It boots the OpenWrt kernel without this change, doesn't it? Patching the bootargs at run-time is the preferred method as it could e.g. cause problems with the stock firmware otherwise.

grosjo commented:
no, as mentioned already so many times, the 3 lines need to be put here :
* bootcmd
* resetcmd
* bootargs

to secure proper booting in openwrt and discarding "tuned" Kuwfi/Yuncore parameters

Note also that your PR is in conflict with master

CodeFetch commented:
@grosjo resetcmd is definitely not needed as this is the command the recovery function would execute if there is no resetcmd defined in the firmware image (as introduced by this PR) and hard-coding something there is unnecessary.
bootcmd is already defined here to first try to boot OpenWrt and if that fails, from the stock firmware address
.
I doubt that bootargs need to be changed. Can you please explain why you think that? What are the original bootargs on your device? As long as the kernel boots everything is good, because then the kernel can patch bootargs and I don't see anything that could be defined in bootargs that could alter the behavior of the kernel boot. Do you understand? bootargs can be patched by the kernel and bootargs have no influence on the kernel booting itself.

In OpenWrt it's
U-Boot starts kernel using bootcmd and gives bootargs as parameters
If patch-cmdline is defined in device support, kernel patches bootargs
Kernel tries to boot

So as far as I know patching bootargs in U-Boot is unnecessary.

grosjo commented:
yes it is, otherewise it does not work 

THis is teh reality, you may complain, but still, this is the reality 

If you do no add the 3 commands, you whole PR is totally useless

CodeFetch commented:
> yes it is, otherewise it does not work

Then there is something wrong with your device support, because patching the kernel command line works well with other devices.

> If you do no add the 3 commands, you whole PR is totally useless

No.

grosjo commented:
> > yes it is, otherewise it does not work
> 
> Then there is something wrong with your device support, because patching the kernel command line works well with other devices.
> 
> > If you do no add the 3 commands, you whole PR is totally useless
> 
> No.

YOU may say No, but the reality is stubborn

THis PR has been made following the long discussion on Kuwfi830(D), and the Kufw830(d) requieres the 3 fields

Beleive it or not, the realisty is the reality

grosjo commented:
and your PR has some issues according to GitHub
![Screenshot from 2019-08-20 20-51-18](https://user-images.githubusercontent.com/2287111/63371197-4a3f2200-c38c-11e9-8a39-edfca46854a2.png)

CodeFetch commented:
> You may say No, but the reality is stubborn

The reality is that the U-Boot bootargs have no influence on the boot process if the kernel command line is being patched and that the resetcmd U-Boot environment variable is only being used if the TFTP image does not contain a resetcmd (which it does with this PR).

> and your PR has some issues according to GitHub

It will be synced when I've tested QCA9563-based YunCore devices and applied the changes requested by pepe2k.

grosjo commented:
Yes bit no

THe device DOES need the 3 command to be saved, that is it..ANything else is superficial

I re-created the setup in anotehr PR, as to make things work, instead of this

CodeFetch commented:
What drugs did you take?

Neustradamus commented:
Any news about this PR?

pepe2k commented:
@CodeFetch I got QCA9563 based YunCore device with similar recovery mode in their U-Boot.
I will assign myself to this PR and will help getting it merged.

Neustradamus commented:
@pepe2k: Nice! :)
What device have you now?

pepe2k commented:
@Neustradamus it's XD4200 and A782.
@CodeFetch I pushed slightly changed version into [my staging tree](https://git.openwrt.org/openwrt/staging/pepe2k.git). I plan also to backport this to 19.07. If you have time, please double check.

Cheers,
Piotr

CodeFetch commented:
@pepe2k Looks good. Thank you.

pepe2k commented:
@CodeFetch merged in 8016f648640e8f73f0dc4fc307fb6bebd12afb88, thanks!

Neustradamus commented:
Thanks a lot @pepe2k and @CodeFetch.
I see a lot of wip in closed section, maybe time to remove it for all?
- https://github.com/openwrt/openwrt/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Aclosed+label%3A"work+in+progress"
